File: 4I505 - CPA/Project/src/Ruby/welzl.rb
require 'benchmark'
require_relative 'geometry'
require_relative 'reading'

def b_md(points)
  # Create the circle.
  if (points.length >= 3)
    Geometry.circumcircle(points[0], points[1], points[2])
  elsif (points.length == 2)
    center = Geometry.middle(points[0], points[1])
    radius = Geometry.distance(points[0], center)
    Geometry::Circle.new(center, radius)
  elsif (points.length == 1)
    Geometry::Circle.new(points[0], 0)
  end
end

def welzl(points, border=[])
  points = points.dup
  border = border.dup
  circle = Geometry.circleNil

  if points.length == 0 or border.length == 3
    circle = b_md(border)
  else
    # Random q in p.
    alea = rand(points.length).to_i
    point = points[alea]
    points.delete_at(alea)

    circle = welzl(points, border)

    if not circle or not circle.contains(point)
      border.push(point)
      circle = welzl(points, border)
    end
  end
  circle
end

tests = Dir.entries(ARGV.first)
tests.shift
tests.shift
File.open "results_welzl.csv", "w" do |fh|
  fh << "x;y;radius;time\n"
  tests.each do |t|
    $stdout << t.to_s << "\r"
    points = read_points t
    circle = nil
    bench = Benchmark.realtime { circle = welzl points }
    fh << circle.center.x << ";" << circle.center.y << ";" << circle.radius << ";" << bench << "\n"
    fh.flush
  end
end

File: 4I505 - CPA/Project/src/C/welzl.c
#include "welzl.h"

// Get a random int.
int nxt_i(int max) {
  return (int)(rand() / (double) RAND_MAX * (max - 1));
}

// Do the b_md calculus.
struct cl b_md(struct pt points[], int size) {
  if (size >= 4)
    printf("Waouh\n");
  if (size == 3) {
    return circum_circle(points[0], points[1], points[2]);
  } else if (size == 2) {
    struct pt center = middle(points[0], points[1]);
    double radius    = distance(points[0], center);
    return new_circle(center, radius);
  } else if (size == 1) {
    return new_circle(points[0], 0);
  }
  return new_circle(new_point(0, 0), -2);
}

// Cause there's an algorithm.
struct cl _welzl(struct pt points[], struct pt border[], size_t size_points, size_t size_border) {
  // Duplicates points.
  {
    struct pt *pts = calloc(512, sizeof *pts);
    struct pt *bor = calloc(512, sizeof *bor);

    memcpy(pts, points, sizeof *pts * size_points);
    memcpy(bor, border, sizeof *bor * size_border);

    points = pts;
    border = bor;
  }

  // Initiliazes the empty circle.
  struct cl circle;

  // And begins.
  if (size_points == 0 || size_border == 3) {
    circle = b_md(border, size_border);
  } else {
    // Pick a random point and remove it from points.
    size_t alea    = nxt_i(size_points);
    struct pt back = points[alea];
    points[alea]   = points[--size_points];

    // Recursive call.
    circle = _welzl(points, border, size_points, size_border);

    // Test if another recursive call is needed.
    if (circle.equal(circle, new_circle(new_point(0, 0), -2)) || !circle.contains(circle, back)) {
      border[size_border++] = back;
      circle = _welzl(points, border, size_points, size_border);
    }
  }

  // And free the memory and return.
  free(points);
  free(border);
  return circle;
}

struct cl welzl(struct pt points[], size_t size) {
  struct pt *start = malloc(sizeof *start * size);
  return _welzl(points, start, size, 0);
}
